### PR TITLE
[TensorExpr] Make sum available from Python

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

File: torch/csrc/jit/ir/constants.cpp [hunk 1]
```diff
@@ -161,6 +161,11 @@ c10::optional<IValue> toIValue(const Value* v) {
     const auto& list = node->ival(attr::value);
     TORCH_INTERNAL_ASSERT(list.isList());
     return list;
+  } else if (
+      type->cast<ListType>() &&
```
Comment (ZolotukhinM) [inline torch/csrc/jit/ir/constants.cpp:165]: Could you please elaborate why this change is needed? Also, this touches one of the foundational files, so it's better to separate this change from everything else and test separately. cc @eellison too, he should know this part better

Comment (asuhan) [reply torch/csrc/jit/ir/constants.cpp:165]: I'm actually not sure about "how needed" this is - for driving this functionality from Python, it's not. However, if I try to parse the generated IR from C++, with `parseIR`, it hits this branch. In other words, looks like we can't parse what we serialize.

I don't know what the best resolution would be here. I'm worried about the fidelity of our IR parser if it creates types we later cannot recognize. I'm also concerned about the fact that we have different ways to represent a list of constant integers. While I believe it's outside the scope of this PR to solve these issues, I think we should fix them in a more principled way.

**Review by ZolotukhinM — COMMENTED** [review]
Good work, I'm super excited to see reductions starting to get fused!

As for the code, it looks to me that there are several independent changes conflated in this PR and it might be a good idea to split them:

1) One part is where we add `aten::sum` to `isSupported` in the fuser pass. This change alone should make sum available from python as we will start fusing sum nodes in the fuser pass.
2) The second part is how we extract dim info from IR. This change doesn't need a fuser support to be implemented and tested - similarly to how the previous PR was done, it could be tested by specifying input IR explicitly.
3) The third part is adjusting what loop transformations we apply in the presence of reductions. I think it's ok to have it as a part of some other change if that's more convenient for you from development perspective, but strictly speaking it's also an independent change.
4) And the last part is the change `constants.cpp` - I am not sure why we need that and whether we need it at all, hopefully Elias could provide more details there.

A reason I brought this all up is that there are several steps in making something available from python and it could actually be better to test each of the steps individually - like we do in `test_te_fuser_pass.cpp` and `test_kernel.cpp`. End to end python tests are of course useful too, but imho should not replace more targeted tests.

File: torch/csrc/jit/ir/constants.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/ir/constants.cpp:165]: Ah, I see. I think I've seen a similar problem before - it's a place in the JIT IR Parser that got slightly stale (originally we used the `is` for lists of ints, then changed to `ival`, but forgot to update the parser). A proper way would probably be to fix the IR parser - I could provide you more details if you're interested in doing that, but I don't know if you'd like to diverge to that work. If not, it's probably better to avoid this change anyway and just try not to hit it in the tests.

**Comment by asuhan** [comment]
Doing just 1 wouldn't help by itself without 2, we're getting the dimensions list in this new format I wasn't aware of - `Constant[value=[0]]()` instead of a `prim::ListConstruct`. IIRC there is an optimization pass which can handle folding the `prim::ListConstruct` version, but I'm not sure how reliable that approach would be or whether such a pass should be mandatory.

We can undo the constants.cpp changes and the associated C++ test change, it's not crucial for this functionality. That'd take care of 4.

For 3, it's just ~2 lines of code. It's not harder for me to separate it, just thought it doesn't warrant an additional commit since it's very small and not easy to test independently either.

File: torch/csrc/jit/ir/constants.cpp (hunk 1 above)
Comment (asuhan) [reply torch/csrc/jit/ir/constants.cpp]: Done, reverted and we'll wait on https://github.com/pytorch/pytorch/pull/43785 to land.

**Comment by asuhan** [comment]
Done, we have 2 as the first commit and 1 + 3 as the second commit in this PR. I don't think either 1 or 3 by itself warrants an individual commit.

**Review by ZolotukhinM — APPROVED** [review]
Looks good! Thanks for addressing my feedback!
